NIFI-3189: ConsumeKafka & Back-pressure. ConsumeKafka_0_10

ijokarumawak opened the pull request:
While downstream connections are full, ConsumeKafka is not scheduled to
run onTrigger, and it won't perform poll, which is important to tell
Kafka server that this client is alive. Thus, after a while in that
situation, Kafka server rebalances the client, and then if downstream
flow backs to normal and ConsumeKafka is scheduled again, the client is no
longer has a valid connection, and a warning message is logged.

This PR uses Kafka pause/resume consumer API to tell Kafka server that
the client doesn't want to consume any message but still alive. It uses
another internal thread to do so by periodically checks when was the
last time that onTrigger is executed. If it has been a while since last
onTrigger (poll), then it pauses the consumer and poll to send a
heartbeat so that the connection will be kept.

Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [x] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

ijokarumawak commented:
Dear reviewers,

This PR contains handful debug level logs that would make easier to understand how the additional thread is working. To see those log messages, modify `conf/logback.xml` as follows:

```xml
<logger name="org.apache.nifi" level="INFO"/>
<logger name="org.apache.nifi.processors" level="WARN"/>
<!-- Add this line -->
<logger name="org.apache.nifi.processors.kafka.pubsub" level="DEBUG" />
<logger name="org.apache.nifi.processors.standard.LogAttribute" level="INFO"/>
<logger name="org.apache.nifi.controller.repository.StandardProcessSession" level="WARN" />
```

This PR is aimed for ConsumeKafka_0_10. Once this is proved to work properly, I'll create another PR for ConsumeKafka (0.9) similarly.

Thanks a lot in advance!

bbende commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-0-10-processors/src/main/java/org/apache/nifi/processors/kafka/pubsub/ConsumerPool.java:
```diff
@@ -126,6 +128,25 @@ public ConsumerLease obtainConsumer(final ProcessSession session) {
     }
 
+    public void retainConsumers() {
+        pooledLeases.forEach(lease -> {
```
Is there any issue if another thread calls obtainConsumer while retainConsumers is in the middle of looping over the consumers?

ijokarumawak commented:
@bbende Thanks for reviewing. Incorporated comment. While I was doing so, I found that Kafka client library started supporting heartbeat in background thread [KAFKA-3888](https://issues.apache.org/jira/browse/KAFKA-3888) since 0.10.1.0.

Maybe we should test the latest Kafka client, that may automatically solve this issue without any modification on NiFi side, although I haven't looked at the new client code yet. If it does, this PR will be purely for ConsumeKafka (0.9).

ijokarumawak commented:
@bbende It turned out to be updating Kafka client is sufficient for 0.10. Confirmed that ConsumeKafka_0_10 can rejoin the consumer group when it resumes after being back-pressure engaged more than 5 minutes. For 0.9, I've added the manual connection retaining logic using paused poll.

- For ConsumeKafka_0_10, use latest client library

    Above issue has been addressed by KIP-62.
    The latest Kafka consumer poll checks if the client instance is still valid, and rejoin the group if not, before consuming messages.

- For ConsumeKafka (0.9), added manual retention logic using pause/resume

    Kafka client 0.9 doesn't have background thread heartbeat, so similar machanism is added manually.
    Use Kafka pause/resume consumer API to tell Kafka server that the client stops consuming messages but is still alive.
    Another internal thread is used to perform paused poll periodically based on the time passed since the last onTrigger(poll) is executed.

bbende commented:
+1 I recreated the scenario with 0.9 and 0.10 by setting back-pressure to 1 on the queues after ConsumeKafka and ConsumeKafka_0_10, and verified that after this patch they have no problem resuming after sitting with back-pressure for long periods of time. Will merge to master.
